instrumentation/jdbi3/src/main/java/brave/jdbi3/Jdbi3BravePlugin.java

```diff
@@ -0,0 +1,173 @@
+                            String bindVariables = ctx.getBinding()
+                                .describe(stmtConfig.getJfrParamMaxLength());
+                            span.tag("sql.binding", bindVariables);
+                        }
+
+                        span.tag("sql.rows", Long.toString(ctx.getMappedRows()));
+
+                        if (ctx.getCompletionMoment() == null) {
+                            span.tag("error", "");
+                        }
+
+                        if (ctx.getCompletionMoment() != null) {
+                            span.finish(ctx.getCompletionMoment().toEpochMilli() * 1000);
+                        } else if (ctx.getExceptionMoment() != null) {
+                            span.finish(ctx.getExceptionMoment().toEpochMilli() * 1000);
+                        } else {
+                            span.finish();
+                        }
+                    });
+                }
+            });
+    }
+
+    static String getRemoteServiceName(StatementContext ctx) {
+        Connection connection = ctx.getConnection();
+        if (connection == null) {
+            return null;
+        }
+        URI uri;
+        try {
+            uri = new URI(connection.getMetaData().getURL().replaceAll("^jdbc:", ""));
+        } catch (SQLException | URISyntaxException ignored) {
+            return null;
+        }
+
+        String remoteServiceNameFromQueryParameter = getRemoteServiceNameFromQueryParameter(uri);
+        if (remoteServiceNameFromQueryParameter != null) {
+            return remoteServiceNameFromQueryParameter;
+        }
+
+        if (uri.getAuthority() != null) {
+            return uri.getAuthority();
+        }
+
+        if (uri.getScheme() != null) {
+            return uri.getScheme();
+        }
+
+        return null;
+    }
+
+    static String getRemoteServiceNameFromQueryParameter(URI uri) {
+        String query = uri.getQuery();
+        if (query != null) {
+            Matcher matcher = serviceNamePattern.matcher(query);
+            if (matcher.find()) {
+                return matcher.group("serviceName");
+            }
+        }
+
+        try {
+            URI schemeSpecificUri = new URI(uri.getSchemeSpecificPart());
+            if (schemeSpecificUri.toString().length() < uri.toString().length()) {
+                return getRemoteServiceNameFromQueryParameter(schemeSpecificUri);
+            }
+        } catch (URISyntaxException ignored) {
+        }
+
+        return null;
+    }
+
+    public static class Builder {
+        private final Tracing tracing;
+        private boolean includeBindVariables = false;
+
+        public Builder(Tracing tracing) {
+            this.tracing = tracing;
+        }
+
+        public Builder includeBindVariables(boolean includeBindVariables) {
+            this.includeBindVariables = includeBindVariables;
+            return this;
+        }
+
+        public Jdbi3BravePlugin build() {
```

Comment:
always return the most narrow type you can that works. Either plugin or plugin.singleton depending on what you think here. This is likely a place where since mysql plugin was made so long ago we didn't make do this.

```suggestion
        public JdbiPlugin.Singleton build() {
```